File: study/ts/script/scene/view/friend/PopupFriendSuggest.ts
const {ccclass, property} = cc._decorator;

import CommonButtonLevel0Highlight from '../../../ui/component/CommonButtonLevel0Highlight'

import CommonNormalLargePop3 from '../../../ui/component/CommonNormalLargePop3'
import PopupBase from '../../../ui/PopupBase';
import UIHelper from '../../../utils/UIHelper';
import { G_ConfigLoader, G_UserData, G_SignalManager, G_ServerTime, G_Prompt } from '../../../init';
import { ConfigNameConst } from '../../../const/ConfigNameConst';
import { assert } from '../../../utils/GlobleFunc';
import { Lang } from '../../../lang/Lang';
import { handler } from '../../../utils/handler';
import { FriendUnitData } from '../../../data/FriendUnitData';
import { SignalConst } from '../../../const/SignalConst';
import { Util } from '../../../utils/Util';
import UIActionHelper from '../../../utils/UIActionHelper';
import CommonCustomListViewEx from '../../../ui/component/CommonCustomListViewEx';
import { FriendConst } from '../../../const/FriendConst';
import { table } from '../../../utils/table';

@ccclass
export default class PopupFriendSuggest extends PopupBase {

    @property({
        type: cc.Node,
        visible: true
    })
    _resourceNode: cc.Node = null;

    @property({
        type: CommonNormalLargePop3,
        visible: true
    })
    _commonLargePop2: CommonNormalLargePop3 = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageRoot: cc.Sprite = null;

    @property({
        type: CommonCustomListViewEx,
        visible: true
    })
    _listView: CommonCustomListViewEx = null;

    @property({
        type: CommonButtonLevel0Highlight,
        visible: true
    })
    _btnRefresh: CommonButtonLevel0Highlight = null;

    @property(cc.Prefab)
    FriendSuggestCell:cc.Prefab = null;

    _countDownStartTime: number = 0;
    _suggestInterval: number;
    _listData: FriendUnitData[];
    _signalSuggestList: any;
    _signalAddFriend: any;

    public static path:string = 'friend/PopupFriendSuggest';
    
    ctor() {
        UIHelper.addEventListener(this.node, this._btnRefresh._button, 'PopupFriendSuggest', '_onBtnRefresh');
    }
    onCreate() {
        this.ctor();
        var Parameter = G_ConfigLoader.getConfig(ConfigNameConst.PARAMETER);//require('app.config.parameter');
        var config = Parameter.get(141);
      //assert((config != null, 'can not find param 141 value');
        this._suggestInterval = parseInt(config.content) || 10;
        this._commonLargePop2.setTitle(Lang.get('lang_friend_suggest_title'));
        this._commonLargePop2.addCloseEventListener(handler(this, this.close));
        this._initListView();
        this._refreshBtnState();
        this._listData = G_UserData.getFriend().getSuggestTempListData();
        this.updateListView();
    }
    onEnter() {
        this._signalSuggestList = G_SignalManager.add(SignalConst.EVENT_RECOMMAND_FRIEND_SUCCESS, handler(this, this._onGetSuggestList));
        this._signalAddFriend = G_SignalManager.add(SignalConst.EVENT_ADD_FRIEND_SUCCESS, handler(this, this._onAddFriendSuccess));
        if (this._countDownStartTime + this._suggestInterval - G_ServerTime.getTime() < 0) {
            this.requestSuggestList();
        }
    }
    onExit() {
        this._signalSuggestList.remove();
        this._signalSuggestList = null;
        this._signalAddFriend.remove();
        this._signalAddFriend = null;
    }
    requestSuggestList() {
        G_UserData.getFriend().c2sRecommandFriend();
    }
    _onGetSuggestList(event, data) {
        this._listData = data;
        this.updateListView();
        this._refreshBtnState();
    }
    _onBtnRefresh() {
        if (this._countDownStartTime == 0) {
            this.requestSuggestList();
        }
    }
    _refreshBtnState() {
        var desc = this._btnRefresh.getDesc();
        if (desc) {
            desc.node.stopAllActions();
            this._countDownStartTime = G_UserData.getFriend().getSuggestRefreshTime();
            if (this._countDownStartTime + this._suggestInterval - G_ServerTime.getTime() >= 0) {
                this._btnRefresh.setEnabled(false);
                this._btnRefresh.setString(Util.format('%ds', this._countDownStartTime + this._suggestInterval - G_ServerTime.getTime()));
                var action = UIActionHelper.createUpdateAction(function () {
                    var leftTime = this._countDownStartTime + this._suggestInterval - G_ServerTime.getTime();
                    if (leftTime > 0) {
                        var desc1 = this._btnRefresh.getDesc();
                        if (desc1) {
                            this._btnRefresh.setString(Util.format('%ds', leftTime));
                        }
                    } else {
                        desc.node.stopAllActions();
                        this._countDownStartTime = 0;
                        this._btnRefresh.setEnabled(true);
                        this._btnRefresh.setString(Lang.get('lang_friend_suggest_btn_refresh'));
                    }
                }.bind(this), 0.5);
                desc.node.runAction(action);
            } else {
                this._btnRefresh.setEnabled(true);
                this._btnRefresh.setString(Lang.get('lang_friend_suggest_btn_refresh'));
            }
        }
    }
    _initListView() {
        this._listView.setTemplate(this.FriendSuggestCell);
        this._listView.setCallback(handler(this, this._onListViewItemUpdate), handler(this, this._onListViewItemSelected));
        this._listView.setCustomCallback(handler(this, this._onListViewItemTouch));
    }
    _onListViewItemUpdate(item, index) {
        if (this._listData) {
            item.updateUI(index + 1, this._listData[index]);
        }
    }
    _onListViewItemSelected(item, index) {
    }
    _onListViewItemTouch(index, data) {
        if (data) {
            G_UserData.getFriend().c2sAddFriend(data.getName(), FriendConst.FRIEND_ADD_FRIEND_TYPE);
        }
    }
    updateListView() {
        if (!this._listData) {
            this._listData = [];
        }
        this._listView.resize(this._listData.length, 2, true);
    }
    _onAddFriendSuccess(event, message) {
        var uid = (message['uid']);
        if (uid && this._listData) {
            for (var k=0; k<this._listData.length; k++) {
                var v = this._listData[k];
                if (v.getId() == uid) {
                    this._listData.splice(k,1);
                    this.updateListView();
                    break;
                }
            }
        }
        var friend_type = (message['friend_type']);
        if (friend_type) {
            if (friend_type == FriendConst.FRIEND_ADD_FRIEND_TYPE) {
                G_Prompt.showTip(Lang.get('lang_friend_apply_success_tip'));
            }
        }
    }

}

File: study/lua/src/app/scene/view/friend/PopupFriendSuggest.lua

-- Author: nieming
-- Date:2017-12-26 17:08:08
-- Describle：

local PopupBase = require("app.ui.PopupBase")
local PopupFriendSuggest = class("PopupFriendSuggest", PopupBase)
local FriendConst = require("app.const.FriendConst")
local UIHelper = require("yoka.utils.UIHelper")
local UIActionHelper = require("app.utils.UIActionHelper")
function PopupFriendSuggest:ctor()

	--csb bind var name
	self._btnRefresh = nil  --CommonButtonHighLight
	self._commonLargePop2 = nil  --CommonNormalSmallPop
	self._imageRoot = nil  --ImageView
	self._listView = nil  --ScrollView

	self._listData = nil

	self._countDownStartTime = 0
	local resource = {
		file = Path.getCSB("PopupFriendSuggest", "friend"),
		binding = {
			_btnRefresh = {
				events = {{event = "touch", method = "_onBtnRefresh"}}
			},
		},
	}
	PopupFriendSuggest.super.ctor(self, resource)
end

-- Describle：
function PopupFriendSuggest:onCreate()

	local Parameter = require("app.config.parameter")
	local config = Parameter.get(141)
	assert(config ~= nil, "can not find param 141 value")
	self._suggestInterval = tonumber(config.content) or 10

	self._commonLargePop2:setTitle(Lang.get("lang_friend_suggest_title"))
	self._commonLargePop2:addCloseEventListener(handler(self, self.close))
	self:_initListView()
	self:_refreshBtnState()
	self._listData = G_UserData:getFriend():getSuggestTempListData()
	self:updateListView()
end

-- Describle：
function PopupFriendSuggest:onEnter()
	self._signalSuggestList = G_SignalManager:add(SignalConst.EVENT_RECOMMAND_FRIEND_SUCCESS, handler(self, self._onGetSuggestList))
	self._signalAddFriend = G_SignalManager:add(SignalConst.EVENT_ADD_FRIEND_SUCCESS, handler(self, self._onAddFriendSuccess))

	if self._countDownStartTime + self._suggestInterval - G_ServerTime:getTime() < 0 then
		self:requestSuggestList()
	end
end

-- Describle：
function PopupFriendSuggest:onExit()
	self._signalSuggestList:remove()
	self._signalSuggestList = nil

	self._signalAddFriend:remove()
	self._signalAddFriend = nil
end

function PopupFriendSuggest:requestSuggestList()
	G_UserData:getFriend():c2sRecommandFriend()
end

function PopupFriendSuggest:_onGetSuggestList(event, data)
	self._listData = data
	self:updateListView()
	self:_refreshBtnState()
end

-- Describle：
function PopupFriendSuggest:_onBtnRefresh()
	-- body
	if self._countDownStartTime == 0 then
		self:requestSuggestList()
	end
end

function PopupFriendSuggest:_refreshBtnState()
	local desc = self._btnRefresh:getDesc()
	if desc then
		desc:stopAllActions()
		self._countDownStartTime = G_UserData:getFriend():getSuggestRefreshTime()
		if self._countDownStartTime + self._suggestInterval - G_ServerTime:getTime() >= 0 then
			self._btnRefresh:setEnabled(false)
			self._btnRefresh:setString(string.format("%ds", self._countDownStartTime + self._suggestInterval - G_ServerTime:getTime()))
			local action = UIActionHelper.createUpdateAction(function()
				local leftTime = self._countDownStartTime + self._suggestInterval - G_ServerTime:getTime()
				if leftTime > 0 then
					local desc = self._btnRefresh:getDesc()
					if desc then
						self._btnRefresh:setString(string.format("%ds", leftTime))
					end
				else
					desc:stopAllActions()
					self._countDownStartTime = 0
					self._btnRefresh:setEnabled(true)
					self._btnRefresh:setString(Lang.get("lang_friend_suggest_btn_refresh"))
				end
			end, 0.5)
			desc:runAction(action)
		else
			self._btnRefresh:setEnabled(true)
			self._btnRefresh:setString(Lang.get("lang_friend_suggest_btn_refresh"))
		end
	end
end

function PopupFriendSuggest:_initListView()
	-- body
	local FriendSuggestCell = require("app.scene.view.friend.FriendSuggestCell")
	self._listView:setTemplate(FriendSuggestCell)
	self._listView:setCallback(handler(self, self._onListViewItemUpdate), handler(self, self._onListViewItemSelected))
	self._listView:setCustomCallback(handler(self, self._onListViewItemTouch))

	-- self._listView:resize()
end

-- Describle：
function PopupFriendSuggest:_onListViewItemUpdate(item, index)
	if self._listData then
		item:updateUI(index+1, self._listData[index+1])
	end
end

-- Describle：
function PopupFriendSuggest:_onListViewItemSelected(item, index)

end

-- Describle：
function PopupFriendSuggest:_onListViewItemTouch(index, data)
	if data then
		G_UserData:getFriend():c2sAddFriend(data:getName(), FriendConst.FRIEND_ADD_FRIEND_TYPE)
	end
end

function PopupFriendSuggest:updateListView()
	if not self._listData then
		self._listData = {}
	end
	self._listView:resize(#self._listData)
end


function PopupFriendSuggest:_onAddFriendSuccess(event, message)

	local uid = rawget(message, "uid")
	if uid and self._listData then
		for k, v in pairs(self._listData) do
			if v:getId() == uid then
				table.remove(self._listData, k)
				self:updateListView()
				break
			end
		end
	end
	local friend_type = rawget(message, "friend_type")
	if friend_type then
		if friend_type == FriendConst.FRIEND_ADD_FRIEND_TYPE then
			G_Prompt:showTip(Lang.get("lang_friend_apply_success_tip"))
		end
	end
end

return PopupFriendSuggest
